args <- commandArgs()

help <- function(){
    cat("PROseq_makeHeatMapsCdt.R :
- Make heatmaps for sense and anti sense strand occupancy and fold change tables generated by PROseq_heatMatrix.R
\tsaved as a cdt file for java TreeView
- This assumes more than one sample including one control and one experiment.
- \n")
    cat("Usage: \n")
    cat("--Dir        : path to heatmap tables                                          [required]\n")
    cat("--Pattern    : grep pattern for heatmap tables to use quotes (ex. PolII.*293T) [optional]\n") 
    cat("--Control    : RData file for control sample; everything before _Tssup.*       [required]\n")    
    cat("--geneLen    : order heatmap cdt file by gene length in regions (0/1)          [default = 0 ]
\t\t if set to zero cdt will be sorted by total occupancy in the window\n")
    cat("--regions    : GenomicRanges object with the transcripts of interest           [required if sorting by length]\n")
    cat("\n")
    q()
}

## Save values of each argument
if(length(args)==0 || !is.na(charmatch("-help",args))){
    help()
} else {
    Dir        <- sub( '--Dir=', '', args[grep('--Dir=', args)])
    Pattern    <- sub( '--Pattern=', '', args[grep('--Pattern=', args)])
    Control    <- sub( '--Control=', '', args[grep('--Control=', args)])
    geneLen    <- sub( '--geneLen=', '',args[grep('--geneLen=',args)])
    regions    <- sub( '--regions=', '', args[grep('--regions=', args)] )
}

#setwd("../../analysis/kevin/SEC/")
#regions <- "tables/Pol2_293T_DMSO_817_rep1.filteredProteinCodingTx_ProDmso.rda"
#Control <- "PRO_DMSO_293T_836.3prime"
#Pattern <- "PRO_DMSO_293T_836.3prime_Tssup2000|PRO_468_293T_836.3prime_Tssup2000|PRO_522_293T_836.3prime_Tssup2000"
#Experiment <- "PRO_468_293T_836_Tssup1000down50000bins50"
#geneLen <- 1
#Dir <- "tables/heatmaps/lengthSort"

Pattern

library(GenomicRanges)

if (identical(geneLen,character(0))){
    geneLen <- 0
}else{
    geneLen <- as.numeric(geneLen)
}

Dir
foo <- list.files(Dir, pattern=".rda", full.names=TRUE)
foo <- foo[grep(Pattern,foo,invert=FALSE)]
foo

for (i in 1:length(foo))
{
    oname = sub("$", ".df", gsub("_Tss.*bins[0-9]*|.rda", "", basename(foo[i])))
    oname <- gsub("-","_",oname)
    df <- get(load(foo[i]))
    colnames(df) <- paste(sub(".df", "", oname), 1:ncol(df), sep=".")
    assign(oname, df)
}

SAMPLES <- ls(pattern=".df$")
SAMPLES <- SAMPLES[grep("^anti.df$|^sense.df$", SAMPLES, invert=TRUE)]
SAMPLES

## sanity check must be the same genes in the same order
stopifnot(rownames(get(SAMPLES[1]))==rownames(get(SAMPLES[3])))

exps.sense <- SAMPLES[grep(paste0(Control, ".*.df"), SAMPLES, invert=TRUE)]
exps.anti  <- exps.sense[grep("anti", exps.sense)]
exps.sense <- exps.sense[grep("anti", exps.sense, invert=TRUE)]
exps.sense
exps.anti

con.sense <- SAMPLES[grep(paste0(Control, ".sense.df"), SAMPLES)]
con.anti  <- SAMPLES[grep(paste0(Control, ".anti.df"), SAMPLES)]
con.sense
con.anti

## report the order in the data frame
df.sense <- get(con.sense)
print(paste("control sample 1", con.sense))
for( i in 1:length(exps.sense) ){
    print(paste("sample", i+1, exps.sense[i]))
    df.sense <- cbind(df.sense, get(exps.sense[i]))
}

df.anti <- get(con.anti)
print(paste("control sample 1", con.anti))
for( i in 1:length(exps.anti) ){
    print(paste("sample", i+1, exps.anti[i]))
    df.anti <- cbind(df.anti, get(exps.anti[i]))
}

fc.sense <- do.call(cbind, lapply(1:length(exps.sense), function(i){
    print(paste("FC sense sample", i, exps.sense[i]))
    exp         <- as.matrix(get(exps.sense[i]))
    con         <- as.matrix(get(con.sense))
    pseudo      <- min(c(min(exp[!exp == 0]), min(con[!con == 0])))
    exp         <- exp + pseudo
    con         <- con + pseudo
    FC          <- log2(exp/con)
    FC
}))

fc.anti <- do.call(cbind, lapply(1:length(exps.anti), function(i){
    print(paste("FC sample", i, exps.anti[i]))
    exp         <- as.matrix(get(exps.anti[i]))
    con         <- as.matrix(get(con.anti))
    pseudo      <- min(c(min(exp[!exp == 0]), min(con[!con == 0])))
    exp         <- exp + pseudo
    con         <- con + pseudo
    FC          <- log2(exp/con)
    FC
}))

## write cdt file for control, experiment and FC for both sense and anti sense
if(geneLen==1){
    ## load geneModel with lengths
    model         <- get(load(regions))
    model$length  <- width(model)
    ## incase you a gene is throw out of the heatmap due to out of bounds be sure that you have the same ones as tables
    model         <- model[names(model) %in% rownames(df.sense)]
    ## sort
    or            <- names( model[order(model$length)]    ) ## get names of genes sorted by length
    sense.cdt     <- cbind( UID=or,NAME=or, df.sense[or,] ) 
    anti.cdt      <- cbind( UID=or,NAME=or, df.anti[or,]  )  
    fc.sense.cdt  <- cbind( UID=or,NAME=or, fc.sense[or,]  )
    fc.anti.cdt   <- cbind( UID=or,NAME=or, fc.anti[or,]  )  
    ## write tables
    write.table(sense.cdt, file=paste(Dir, paste0(Control, ".lenSort.cov.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(anti.cdt,  file=paste(Dir, paste0(Control, ".lenSort.cov.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    write.table(fc.sense.cdt, file=paste(Dir, paste0(Control, ".lenSort.log2FC.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(fc.anti.cdt,  file=paste(Dir, paste0(Control, ".lenSort.log2FC.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
}else{
    ## sort
    or            <- order(-rowSums(get(con.sense)))
    sense.cdt     <- cbind( UID=or,NAME=or, df.sense[or,] ) 
    anti.cdt      <- cbind( UID=or,NAME=or, df.anti[or,]  )  
    fc.sense.cdt  <- cbind( UID=or,NAME=or, fc.sense[or,]  )
    fc.anti.cdt   <- cbind( UID=or,NAME=or, fc.anti[or,]  )  
    ## write tables
    write.table(sense.cdt, file=paste(Dir, paste0(Control, ".ocpSort.cov.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(anti.cdt,  file=paste(Dir, paste0(Control, ".ocpSort.cov.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    write.table(fc.sense.cdt, file=paste(Dir, paste0(Control, ".ocpSort.log2FC.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(fc.anti.cdt,  file=paste(Dir, paste0(Control, ".ocpSort.log2FC.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
}
