args <- commandArgs()

help <- function(){
    cat("PROseq_makeHeatMapsCdt.R :
- Make heatmaps for sense and anti sense strand occupancy and fold change tables generated by PROseq_heatMatrix.R
saved as a cdt file for java TreeView
- \n")
    cat("Usage: \n")
    cat("--Dir        : path to heatmap tables                                          [required]\n")
    cat("--Control    : RData file for control sample                                   [required]\n")    
    cat("--Experiment : RData file for experimental sample                              [required]\n")    
    cat("--geneLen    : order heatmap cdt file by gene length in regions (0/1)          [default = 0 ]
\t\t if set to zero cdt will be sorted by total occupancy in the window\n")
    cat("--regions    : GenomicRanges object with the transcripts of interest           [required if sorting by length]\n")
    cat("\n")
    q()
}

## Save values of each argument
if(length(args)==0 || !is.na(charmatch("-help",args))){
    help()
} else {
    Dir        <- sub( '--Dir=', '', args[grep('--Dir=', args)])
    Control    <- sub( '--Control=', '', args[grep('--Control=', args)])
    Experiment <- sub( '--Experiment=', '', args[grep('--Experiment=', args)])
    geneLen    <- sub( '--geneLen=', '',args[grep('--geneLen=',args)])
    regions    <- sub( '--regions=', '', args[grep('--regions=', args)] )
}

## for testing be sure to comment out!
#regions    <- "tables/PRO_0h_Aux_PAF1AID_DLD1_846tss500.filteredProteinCodingTx.rda"
#Control    <- "PRO_0h_Aux_PAF1AID_DLD1_846_Tssup1000down50000bins50"
#Experiment <- "PRO_1D_Aux_PAF1AID_DLD1_846_Tssup1000down50000bins50"
#Dir        <- "tables/heatmaps"
#geneLen    <- 0
#setwd("../")
setwd("../../analysis/kevin/SEC/")

regions <- "tables/Pol2_293T_DMSO_817_rep1.filteredProteinCodingTx_ProDmso.rda"
Control <- "PRO_DMSO_293T_836.36bp_Tssup1000down50000bins50"
Experiment <- "PRO_468_293T_836_Tssup1000down50000bins50"
geneLen <- 1
Dir <- "tables/heatmaps/lengthSort"

library(GenomicRanges)

if (identical(geneLen,character(0))){
    geneLen <- 0
}else{
    geneLen <- as.numeric(geneLen)
}

## load files and assign to file name
con.anti  <- get(load(paste(Dir, paste0(Control, ".anti.rda"),     sep="/")))
con.sense <- get(load(paste(Dir, paste0(Control, ".sense.rda"),    sep="/")))
exp.anti  <- get(load(paste(Dir, paste0(Experiment, ".anti.rda"),  sep="/")))
exp.sense <- get(load(paste(Dir, paste0(Experiment, ".sense.rda"), sep="/")))

## write cdt file for control, experiment and FC for both sense and anti sense
if(geneLen==1){
    ## load geneModel with lengths
    model         <- get(load(regions))
    model$length  <- width(model)
    ## incase you a gene is throw out of the heatmap due to out of bounds be sure that you have the same ones as tables
    model         <- model[names(model) %in% rownames(con.sense)]
    ## sort
    or            <- names( model[order(model$length)]    ) ## get names of genes sorted by length
    con.sense.cdt <- cbind( UID=or,NAME=or, con.sense[or,] ) 
    con.anti.cdt  <- cbind( UID=or,NAME=or, con.anti[or,]  )  
    exp.sense.cdt <- cbind( UID=or,NAME=or, exp.sense[or,] ) 
    exp.anti.cdt  <- cbind( UID=or,NAME=or, exp.anti[or,]  )  
    ## write tables
    write.table(con.sense.cdt, file=paste(Dir, paste0(Control, ".lenSort.cov.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(con.anti.cdt,  file=paste(Dir, paste0(Control, ".lenSort.cov.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    write.table(exp.sense.cdt, file=paste(Dir, paste0(Experiment, ".lenSort.cov.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(exp.anti.cdt,  file=paste(Dir, paste0(Experiment, ".lenSort.cov.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    ## FC heatmap sense
    FC.sense                        <- as.matrix(exp.sense[or,] / con.sense[or,])
    FC.sense[is.nan(FC.sense)]      <- 1
    FC.sense[is.infinite(FC.sense)] <- max(FC.sense[!FC.sense %in% "Inf"])
    FC.sense                        <- log2(FC.sense)
    FC.sense[is.infinite(FC.sense)] <- min(FC.sense[!FC.sense %in% "-Inf"])
    write.table(cbind(UID=rownames(FC.sense)
                     ,NAME=rownames(FC.sense),FC.sense)
               ,file=paste(Dir, paste0(Experiment,"_vs_",Control, ".lenSort.log2FC.sense.cdt"),  sep="/"), sep="\t",row.names=FALSE)

    ## FC heatmap anti
    FC.anti                       <- as.matrix(exp.anti[or,] / con.anti[or,])
    FC.anti[is.nan(FC.anti)]      <- 1
    FC.anti[is.infinite(FC.anti)] <- max(FC.anti[!FC.anti %in% "Inf"])
    FC.anti                       <- log2(FC.anti)
    FC.anti[is.infinite(FC.anti)] <- min(FC.anti[!FC.anti %in% "-Inf"])
    write.table(cbind(UID=rownames(FC.anti)
                     ,NAME=rownames(FC.anti),FC.anti)
               ,file=paste(Dir, paste0(Experiment,"_vs_",Control, ".lenSort.log2FC.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
}else{
    ## sort
    or            <- order(-rowSums(con.sense))
    con.sense.cdt <- cbind( UID=or,NAME=or, con.sense[or,] ) 
    con.anti.cdt  <- cbind( UID=or,NAME=or, con.anti[or,]  )  
    exp.sense.cdt <- cbind( UID=or,NAME=or, exp.sense[or,] ) 
    exp.anti.cdt  <- cbind( UID=or,NAME=or, exp.anti[or,]  )  
    ## write tables
    write.table(con.sense.cdt, file=paste(Dir, paste0(Control, ".ocpSort.cov.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(con.anti.cdt,  file=paste(Dir, paste0(Control, ".ocpSort.cov.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    write.table(exp.sense.cdt, file=paste(Dir, paste0(Experiment, ".ocpSort.cov.sense.cdt"), sep="/"), sep="\t",row.names=FALSE)
    write.table(exp.anti.cdt,  file=paste(Dir, paste0(Experiment, ".ocpSort.cov.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    ## FC heatmap multi
    FC.sense                        <- as.matrix(exp.sense[or,] / con.sense[or,])
    FC.sense[is.nan(FC.sense)]      <- 1
    FC.sense[is.infinite(FC.sense)] <- max(FC.sense[!FC.sense %in% "Inf"])
    FC.sense                        <- log2(FC.sense)
    FC.sense[is.infinite(FC.sense)] <- min(FC.sense[!FC.sense %in% "-Inf"])
    write.table(cbind(UID=rownames(FC.sense),NAME=rownames(FC.sense),FC.sense)
               ,file=paste(Dir, paste0(Experiment,"_vs_",Control, ".ocpSort.log2FC.sense.cdt"),  sep="/"), sep="\t",row.names=FALSE)
    ## FC heatmap anti
    FC.anti                       <- as.matrix(exp.anti[or,] / con.anti[or,])
    FC.anti[is.nan(FC.anti)]      <- 1
    FC.anti[is.infinite(FC.anti)] <- max(FC.anti[!FC.anti %in% "Inf"])
    FC.anti                       <- log2(FC.anti)
    FC.anti[is.infinite(FC.anti)] <- min(FC.anti[!FC.anti %in% "-Inf"])
    ## cdt format requires UID and NAME column
    write.table(cbind(UID=rownames(FC.anti)
                     ,NAME=rownames(FC.anti),FC.anti)
               ,file=paste(Dir, paste0(Experiment,"_vs_",Control, ".ocpSort.log2FC.anti.cdt"),  sep="/"), sep="\t",row.names=FALSE)
}
